src/query/service/src/servers/flight/flight_client.rs

```diff
@@ -225,54 +290,159 @@ impl Drop for FlightReceiver {
 }
 
 impl FlightReceiver {
-    pub fn create(rx: Receiver<Result<FlightData>>) -> FlightReceiver {
+    pub fn create(
+        rx: Receiver<Result<FlightData>>,
+        connection_info: Option<ConnectionInfo>,
+    ) -> FlightReceiver {
         FlightReceiver {
-            rx,
-            notify: Arc::new(WatchNotify::new()),
+            seq_num: AtomicUsize::new(0),
+            rx: RwLock::new(rx),
+            notify: RwLock::new(Arc::new(WatchNotify::new())),
+            connection_info,
         }
     }
 
     #[async_backtrace::framed]
     pub async fn recv(&self) -> Result<Option<DataPacket>> {
-        match self.rx.recv().await {
+        let receiver = self.rx.read().clone();
```

Comment:
```suggestion
        let receiver = self.rx.read();
```

Reply:
It will raise a compile error: we cannot hold the lock when we use await.